libs/community/langchain_community/vectorstores/lancedb.py

```diff
@@ -227,60 +512,118 @@ def similarity_search(
-                    for col in columns
-                    if col != self._text_key
-                },
+        if k is None:
+            k = self.limit
+
+        if self._embedding is None:
+            raise ValueError(
+                "For MMR search, you must specify an embedding function on" "creation."
             )
-            for idx in range(len(docs))
-        ]
+
+        embedding = self._embedding.embed_query(query)
+        docs = self.max_marginal_relevance_search_by_vector(
+            embedding,
+            k,
+            fetch_k,
+            lambda_mult=lambda_mult,
+            filter=filter,
+        )
+        return docs
+
+    def max_marginal_relevance_search_by_vector(
+        self,
+        embedding: List[float],
+        k: Optional[int] = None,
+        fetch_k: int = 20,
+        lambda_mult: float = 0.5,
+        filter: Optional[Dict[str, str]] = None,
+        **kwargs: Any,
+    ) -> List[Document]:
+        """Return docs selected using the maximal marginal relevance.
+        Maximal marginal relevance optimizes for similarity to query AND diversity
+        among selected documents.
+
+        Args:
+            embedding: Embedding to look up documents similar to.
+            k: Number of Documents to return. Defaults to 4.
+            fetch_k: Number of Documents to fetch to pass to MMR algorithm.
+            lambda_mult: Number between 0 and 1 that determines the degree
+                        of diversity among the results with 0 corresponding
+                        to maximum diversity and 1 to minimum diversity.
+                        Defaults to 0.5.
+            filter (Optional[Dict[str, str]]): Filter by metadata. Defaults to None.
+
+        Returns:
+            List of Documents selected by maximal marginal relevance.
+        """
+
+        results = self._query(
+            query=embedding,
+            k=fetch_k,
+            filter=filter,
+            **kwargs,
+        )
+        mmr_selected = maximal_marginal_relevance(
+            np.array(embedding, dtype=np.float32),
+            results["vector"].to_pylist(),
+            k=k or self.limit,
+            lambda_mult=lambda_mult,
+        )
+
+        candidates = self.results_to_docs(results)
+
+        selected_results = [r for i, r in enumerate(candidates) if i in mmr_selected]
+        return selected_results
 
     @classmethod
     def from_texts(
-        cls,
+        cls: Type[LanceDB],
         texts: List[str],
         embedding: Embeddings,
         metadatas: Optional[List[dict]] = None,
-        connection: Any = None,
+        connection: Optional[Any] = None,
         vector_key: Optional[str] = "vector",
         id_key: Optional[str] = "id",
         text_key: Optional[str] = "text",
         table_name: Optional[str] = "vectorstore",
+        api_key: Optional[str] = None,
+        region: Optional[str] = None,
+        mode: Optional[str] = "overwrite",
+        distance: Optional[str] = "l2",
```

Comment:
Do these need to be optional?

Reply:
The api_key, region would be optional as they are only if user wants to use cloud, I change the mode/distance args to be fixed